### PR TITLE
Register DependencyContextRazorViewEngineOptionsSetup after RazorViewEngineOptionsSetup

### LINKED ISSUE #4902: OptimizationLevel specified in the dependency context \ primary assembly's project.json is never used when compiling views
`DependencyContextViewEngineOptionsSetup` and `RazorViewEngineOptionsSetup` modify compilation options - the first based on what's specified in the dependency context and second based on `IHostingEnviornment.EnvironmentName`. However the former setup's option values are always overwritten by `RazorViewEngineOptionsSetup` which runs second.

Additionally, the value set in `OptimizationLevel` determined in `DependencyContextViewEngineOptionsSetup` is incorrect - https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/Internal/DependencyContextRazorViewEngineOptionsSetup.cs#L60-L62 needs to be inverted so that when your project.json says `optimize`, we choose `Release`.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #4902

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs [hunk 1]
```diff
@@ -127,18 +127,17 @@ public static IMvcCoreBuilder AddTagHelpersAsServices(this IMvcCoreBuilder build
...
             services.TryAddSingleton<ICompilationService, DefaultRoslynCompilationService>();
 
             services.TryAddEnumerable(
                 ServiceDescriptor.Transient<IConfigureOptions<MvcViewOptions>, MvcRazorMvcViewOptionsSetup>());
             services.TryAddEnumerable(
                 ServiceDescriptor.Transient<IConfigureOptions<RazorViewEngineOptions>, RazorViewEngineOptionsSetup>());
+            services.TryAddEnumerable(
+                ServiceDescriptor.Transient<
+                    IConfigureOptions<RazorViewEngineOptions>,
+                    DependencyContextRazorViewEngineOptionsSetup>());
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs]: We should really have a comment about the required ordering of these.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DependencyContextRazorViewEngineOptionsSetup.cs [hunk 2]
```diff
@@ -10,34 +10,54 @@
...
+using DependencyContextOptions = Microsoft.Extensions.DependencyModel.CompilationOptions;
 
 namespace Microsoft.AspNetCore.Mvc.Razor.Internal
 {
     /// <summary>
     /// Sets up compilation and parse option default options for <see cref="RazorViewEngineOptions"/> using
     /// <see cref="DependencyContext"/>
     /// </summary>
-    public class DependencyContextRazorViewEngineOptionsSetup : ConfigureOptions<RazorViewEngineOptions>
+    public class DependencyContextRazorViewEngineOptionsSetup : IConfigureOptions<RazorViewEngineOptions>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DependencyContextRazorViewEngineOptionsSetup.cs]: :heart_eyes_cat:

File: src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngineOptionsSetup.cs [hunk 3]
```diff
@@ -12,51 +12,51 @@ namespace Microsoft.AspNetCore.Mvc
     /// <summary>
     /// Sets up default options for <see cref="RazorViewEngineOptions"/>.
     /// </summary>
-    public class RazorViewEngineOptionsSetup : ConfigureOptions<RazorViewEngineOptions>
+    public class RazorViewEngineOptionsSetup : IConfigureOptions<RazorViewEngineOptions>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngineOptionsSetup.cs]: Breaking change to a `public` class in `Microsoft.AspNetCore.Mvc`.

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/RazorViewEngineOptionsSetupTest.cs [hunk 4]
```diff
@@ -2,13 +2,12 @@
...
 
 using Microsoft.AspNetCore.Hosting;
-using Microsoft.AspNetCore.Mvc.Razor;
 using Microsoft.CodeAnalysis;
 using Microsoft.Extensions.FileProviders;
 using Moq;
 using Xunit;
 
-namespace Microsoft.AspNetCore.Mvc
+namespace Microsoft.AspNetCore.Mvc.Razor
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/RazorViewEngineOptionsSetupTest.cs]: Namespace and directory were correct before.

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/project.json [hunk 5]
```diff
@@ -10,7 +10,10 @@
         "../Microsoft.AspNetCore.Mvc.Razor.Host.Test/TestFileTrigger.cs",
         "../Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/TestViewBufferScope.cs"
       ]
-    }
+    },
+    "define": [
+      "SOME_TEST_DEFINE"
+    ]
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/project.json]: Suggest `#define SOME_TEST_DEFINE` in the test class to keep it all together.

**Comment by pranavkm** [comment]
:up: :date:

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/DependencyContextRazorViewEngineOptionsSetupTest.cs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+            var setup = new TestableDependencyContextOptionsSetup(hostingEnvironment.Object);
+
+            // Act
+            var options = setup.GetCompilationOptionsPublic();
+
+            // Assert
+            Assert.Same(DependencyContextOptions.Default, options);
+        }
+
+#if SOME_TEST_DEFINE
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/DependencyContextRazorViewEngineOptionsSetupTest.cs]: Don't do this. Want the test to fail if someone removes the definition from the `project.json` file. (My suggestion was `#define` not `#if`.)

**Comment by dougbu** [comment]
:watch: for https://github.com/aspnet/Mvc/pull/4987#discussion-diff-70921004 resolution.

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/DependencyContextRazorViewEngineOptionsSetupTest.cs (hunk 6 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/DependencyContextRazorViewEngineOptionsSetupTest.cs]: Sorry, I don't entirely follow what the `#define` does in this case.

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/project.json (hunk 5 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/project.json]: Doesn't work unfortunately:

```
Not found: SOME_TEST_DEFINE
      In value:  String[] ["DEBUG", "TRACE", "NETCOREAPP1_0"]
```

**Comment by pranavkm** [comment]
Removed the ifdef.

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/project.json (hunk 5 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/project.json]: Oh well. Thx for trying.

**Comment by dougbu** [comment]
:shipit:

**Pull request closed without merging**